x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java

```diff
@@ -596,53 +626,77 @@ public void onFileChanged(Path file) {
     }
 
     /**
-     * Creates a new {@link PrivilegedResourceRetriever} to be used with the {@link IDTokenValidator} by passing the
-     * necessary client SSLContext and hostname verification configuration
+     * Remote JSON Web Key source specified by a JWKSet URL. The retrieved JWK set is cached to
+     * avoid unnecessary http requests. A single attempt to update the cached set is made
+     * (with {@ling ReloadableJWKSource#triggerReload}) when the {@link IDTokenValidator} fails
+     * to validate an ID Token (because of an unknown key) as this might mean that the OpenID
+     * Connect Provider has rotated the signing keys.
      */
-    private PrivilegedResourceRetriever getPrivilegedResourceRetriever() {
-        final String sslKey = RealmSettings.realmSslPrefix(realmConfig.identifier());
-        final SSLConfiguration sslConfiguration = sslService.getSSLConfiguration(sslKey);
-        boolean isHostnameVerificationEnabled = sslConfiguration.verificationMode().isHostnameVerificationEnabled();
-        final HostnameVerifier verifier = isHostnameVerificationEnabled ? new DefaultHostnameVerifier() : NoopHostnameVerifier.INSTANCE;
-        return new PrivilegedResourceRetriever(sslService.sslContext(sslConfiguration), verifier, realmConfig);
-    }
+    class ReloadableJWKSource<C extends SecurityContext> implements JWKSource<C> {
 
-    static class PrivilegedResourceRetriever extends DefaultResourceRetriever {
-        private SSLContext clientContext;
-        private HostnameVerifier verifier;
-        private RealmConfig config;
+        private volatile JWKSet cachedJwkSet = null;
```

Comment:
can we use `new JWKSet()` instead of null? Otherwise there is a chance we get a NPE in the `get` method. If you prefer null, then the get method should be something like:

```java
JWKSet local = this.jwkSet;
if (local != null) {
    return jwkSelector.select(local);
} else {
    return Collections.emptyList();
}
```

Reply:
I will change it but I think null is fine too because `JWKSelector#select` explicitly handles it
```
	/**
	 * Selects the keys from the specified JWK set according to the
	 * matcher's criteria.
	 *
	 * @param jwkSet The JWK set. May be {@code null}.
	 *
	 * @return The selected keys, ordered by their position in the JWK set,
	 *         empty list if none were matched or the JWK is {@code null}.
	 */
```